### PR TITLE
ARROW-9163: [C++] Validate UTF8 contents of a StringArray

### PR DESCRIPTION
**Pull request by pitrou**
* Add a ValidateUTF8() method to StringArray and LargeStringArray
* Automatically call ValidateUTF8() when ValidateFull() is called on
  one of those types

### CODE REVIEW

**Comment by kiszk** [comment]
Looks good

**Comment by pitrou** [comment]
Rebased.

File: cpp/src/arrow/array/array_binary.cc [hunk 1]
```diff
@@ -24,11 +24,27 @@
...
+template <typename StringArrayType>
+Status ValidateStringData(const StringArrayType& array) {
+  util::InitializeUTF8();
+  for (int64_t i = 0; i < array.length(); ++i) {
+    if (!array.IsNull(i) && !util::ValidateUTF8(array.GetView(i))) {
+      return Status::Invalid("Invalid UTF8 sequence at string index ", i);
+    }
+  }
+  return Status::OK();
+}
```
Comment (wesm) [inline cpp/src/arrow/array/array_binary.cc:44]: If this became a perf bottleneck, this would be pretty easy to speed up

**Review by wesm — APPROVED** [review]
+1. There's some occasional naming inconsistency re: `UTF8` versus `Utf8`. Note that the GSG would recommend `Utf8` over `UTF8`, just as `Url` over `URL`. This isn't a big deal just pointing it out

**Pull request closed without merging**